src/Aquifer.Common/Services/CachingLanguageService.cs

```diff
@@ -0,0 +1,75 @@
+
+namespace Aquifer.Common.Services;
+
+public interface ICachingLanguageService
+{
+    Task<string?> GetLanguageCodeAsync(int languageId, CancellationToken ct);
+    Task<IReadOnlyDictionary<int, string>> GetLanguageCodeByIdMapAsync(CancellationToken ct);
+    Task<int?> GetLanguageIdAsync(string languageCode, CancellationToken ct);
+    Task<IReadOnlyDictionary<string, int>> GetLanguageIdByCodeMapAsync(CancellationToken ct);
+}
+
+public class CachingLanguageService(AquiferDbContext _dbContext, IMemoryCache _memoryCache) : ICachingLanguageService
+{
+    public async Task<string?> GetLanguageCodeAsync(int languageId, CancellationToken ct)
+    {
+        return (await GetLanguageCodeByIdMapAsync(ct))
+            .TryGetValue(languageId, out var languageCode)
+                ? languageCode
+                : null;
+    }
+
+    public async Task<IReadOnlyDictionary<int, string>> GetLanguageCodeByIdMapAsync(CancellationToken ct)
+    {
+        return (await GetDictionariesFromCacheAsync(ct))
+            .LanguageCodeByIdMap;
+    }
+
+    public async Task<int?> GetLanguageIdAsync(string languageCode, CancellationToken ct)
+    {
+        ArgumentNullException.ThrowIfNull(languageCode);
+
+        return (await GetLanguageIdByCodeMapAsync(ct))
+            .TryGetValue(languageCode, out var languageId)
+                ? languageId
+                : null;
+    }
+
+    public async Task<IReadOnlyDictionary<string, int>> GetLanguageIdByCodeMapAsync(CancellationToken ct)
+    {
+        return (await GetDictionariesFromCacheAsync(ct))
+            .LanguageIdByCodeMap;
+    }
+
+    private async Task<(IReadOnlyDictionary<int, string> LanguageCodeByIdMap, IReadOnlyDictionary<string, int> LanguageIdByCodeMap)> GetDictionariesFromCacheAsync(CancellationToken ct)
+    {
+        return await _memoryCache.GetOrCreateAsync(
+            LanguageDictionariesCacheKey,
+            async cacheEntry =>
+            {
+                cacheEntry.SlidingExpiration = s_cacheLifetime;
+
+                var languageIdentifiers = await GetLanguageIdentifiersFromDatabaseAsync(_dbContext, ct);
+
+                return
+                    (LanguageCodeById: languageIdentifiers.ToDictionary(li => li.Id, li => li.Code),
+                    LanguageIdByCode: languageIdentifiers.ToDictionary(li => li.Code, li => li.Id));
+            });
+
+        static async Task<IReadOnlyList<LanguageIdentifier>> GetLanguageIdentifiersFromDatabaseAsync(AquiferDbContext dbContext, CancellationToken ct)
+        {
+            return await dbContext.Languages
+                .Select(l => new LanguageIdentifier(l.Id, l.ISO6393Code))
+                .ToListAsync(ct);
+        }
+    }
+
+    private record LanguageIdentifier(int Id, string Code);
+
+    private const string LanguageDictionariesCacheKey = $"{nameof(CachingLanguageService)}:LanguageDictionaries";
+
+    private static readonly TimeSpan s_cacheLifetime = TimeSpan.FromMinutes(30);
```

Comment:
I generally see private variables like this at the top of a file (I think it's helpful for a user to quickly see what defaults are being set for the rest of the file). Thoughts?

Reply:
I'm fine implementing whatever practice we prefer, though I might accidentally lapse into old habits if we don't have a compiler or linter enforcing it.

Some Google searching tells me that there aren't any documented C# framework guidelines for best practices:
* https://www.reddit.com/r/csharp/comments/16zihxz/order_of_methods_in_class/
* https://stackoverflow.com/questions/150479/order-of-items-in-classes-fields-properties-constructors-methods/27997842

We could probably get some enforcement from `.editorconfig` and I know that StyleCop can allow enforcing a certain order because I've used that previously.

My reasoning for `private` things being last is summed up in [this answer](https://stackoverflow.com/a/27997842).  I was initially compelled into `private` fields being at the end of the class when I started working at my last job and didn't like it at first because I initially learned with C++ style declarations that _required_ fields to be declared before they were used (and thus they had to be at the top of a class).  I think the `private fields first` is a holdover from that old compiler requirement.  C# allows any ordering and now I see value in putting the constructors and other methods at the top (with `public` ones first) because that's what I tend to want to read first and the private fields are implementation details of those methods (so I want to read them last).  Code is read more than it is written so defaulting to an ordering that helps the reading process helps with understandability.

Reply:
Yeah, I definitely like seeing the constructor first, and the primary constructors in general handle that now. My format on save throws them up top, and that's how I've always done it, so I don't think much of it. Throughout the project I think they're mostly up top now. I don't have a strong preference, just wanted to throw it out there since I know some of our IDEs are gonna move them automatically.

Reply:
I moved the private fields to the top.